### PR TITLE
New share provider: Share by mail

### LINKED ISSUE #1703: [Feature] Sharing with external email ( not federated )
### Steps to reproduce
1. type an email address with a domain not federated like user@gmail.com
### Expected behaviour

User expect that a custom share link would be generated and sent to this email, later on , the share could be deactivated so this link would not work anymore.
### Actual behaviour

Error is returned : federated domain cannot be reach
### Server configuration

**Operating system**: Any

**Web server:** Any

**Database:** Any

**PHP version:** Any

**Nextcloud version:** 10.0.1

**Updated from an older Nextcloud/ownCloud or fresh install:** NO

**Where did you install Nextcloud from:** zip file

</details>

**Are you using external storage : NO 

**Are you using encryption:** yes/no

**Are you using an external user-backend,  : NO

Already discussed here : 

https://github.com/owncloud/core/issues/20074

### PR DESCRIPTION
**Pull request by schiessle**
This introduces a new share provider called "Share by mail" which allows you to enter a email address to the share dialog and send a personalized link.

share-by-mail

This is work in progress, but at least the auto-complete function works already.

Fix #1703

### CODE REVIEW

**Comment by schiessle** [comment]
@jancborchardt I'm sure that's also interesting to you :wink:

**Comment by nickvergessen** [comment]
missing from shipped.json

**Comment by rullzer** [comment]
While I think the idea is good. The UX there is terrible. Basically it should fall back to sending the e-mail if no remote is found.

File: apps/sharebymail/lib/FederatedShareProvider.php [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+		$shareWith = $share->getSharedWith();
+		$itemSource = $share->getNodeId();
+		$itemType = $share->getNodeType();
+		$permissions = $share->getPermissions();
+		$sharedBy = $share->getSharedBy();
+
+		/*
+		 * Check if file is not already shared with the remote user
+		 */
+		$alreadyShared = $this->getSharedWith($shareWith, self::SHARE_TYPE_REMOTE, $share->getNode(), 1, 0);
```
Comment (icewind1991) [inline apps/sharebymail/lib/FederatedShareProvider.php]: `SHARE_TYPE_EMAIL`?

File: apps/sharebymail/lib/FederatedShareProvider.php [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+			throw new \Exception($message_t);
+		}
+
+
+		// don't allow federated shares if source and target server are the same
+		list($user, $remote) = $this->addressHandler->splitUserRemote($shareWith);
+		$currentServer = $this->addressHandler->generateRemoteURL();
+		$currentUser = $sharedBy;
+		if ($this->addressHandler->compareAddresses($user, $remote, $currentUser, $currentServer)) {
+			$message = 'Not allowed to create a federated share with the same user.';
```
Comment (icewind1991) [inline apps/sharebymail/lib/FederatedShareProvider.php]: email share, not federated share

**Comment by icewind1991** [comment]
The share provider looks like a lot of duplicate code from the other share providers, probably worth it having a go at factoring out the common logic

**Comment by jancborchardt** [comment]
@schiessle awesome! :) Great that you’re pushing this!

I tend to agree with @rullzer – it’s a bit strange at the moment that you need to decide between remote and email. It should check automatically if a remote by that ID exists – if not share via email.

Also, can we tune that dropdown so that to the left of the remote user it shows their avatar (or colored circle with initial) and next to email it shows an email icon? Or do we need to do a major rewire for that?

**Comment by schiessle** [comment]
> The share provider looks like a lot of duplicate code from the other share providers, probably worth it having a go at factoring out the common logic

yes, that's mainly copy&pased from the federated share provider... This will change a lot, you can ignore it for now

**Comment by schiessle** [comment]
> I tend to agree with @rullzer – it’s a bit strange at the moment that you need to decide between remote and email. It should check automatically if a remote by that ID exists – if not share via email.

I'm not sure. This would mean that you would first hand the share to the federated share provider, which would try to create a share. If it fails it would return to the share manager which would then try to do a mail share. Such a behavior is also completely in-transparent to the user. I tend to let the user explicitly decide what he want to do. In case we find a exact match of a federated share or a email address we can hide the other option because it is really unlikely that a user has the exact same email address and federated cloud id. But if we enter something for the first time I would just let the user chose.

> Also, can we tune that dropdown so that to the left of the remote user it shows their avatar (or colored circle with initial) and next to email it shows an email icon? Or do we need to do a major rewire for that?

We don't have a avatar if your share with a user for the first time... Let's keep this for later. This is out of scope for this PR.

**Comment by jancborchardt** [comment]
> I tend to let the user explicitly decide what he want to do.

Nextcloud tends to do stuff automatically. ;) If you share something with me via email – either cause my server is not available or you don’t have my federated cloud ID – it could be added to my Nextcloud as a federated share and communicated back to your server like that. Just automatically, there should be no difference and there doesn’t need to be, because in both cases you have the control over if I »can edit« or not.

**Comment by mar1u5** [comment]
Uhmm ... Isnt this covered by https://github.com/owncloud/core/25576

**Comment by schiessle** [comment]
> Nextcloud tends to do stuff automatically. ;) If you share something with me via email – either cause my server is not available or you don’t have my federated cloud ID – it could be added to my Nextcloud as a federated share and communicated back to your server like that. 

That's confusing. Sounds to me like you mixed federated shares and mail shares up. :wink: 

Let me try to illustrate the two possible scenarios if we try to make it automatically:

1. We first try to do a federated share. If it succeed fine. But what do we do if it don't succeed? If we know it is a federated share ID we can retry it and the recipient will receive his share once his Nextcloud is back online -> everything is fine. If we try to be more intelligent as the user we would fall back to a mail share, we would send the mail to the cloud ID. The delivery will fail because in most cases the cloud ID is not a email address at the same time. We will have no chance to detect that sending the mail failed. The user who send the share will see it in his share list and assume that everything is fine because he entered a federated cloud id. -> this is a problem you can't solve.

2. So we could try to do it the other way around, first try to send a mail.  If it is a valid mail address everything fine. If the rare case happens (0.1%) that a user has the same federated cloud ID as email address this would be bad because he would never receive a federated share but always a email. But it is just 0.1% so let's ignore it for now. Let's assume the ID was a cloud ID, we tried to send a mail and will not have any response if the mail was delivered successfully or not, so we never know if we should retry a federated share -> solution is not feasible at all.

So we have two solutions to automate it and both don't work.

**Comment by schiessle** [comment]
> Uhmm ... Isnt this covered by https://github.com/owncloud/core/25576

@mar1u5 link doesn't work

**Comment by mar1u5** [comment]
Was updated 1 minute later :unamused: ... The /pull/ was missing :wink:

**Comment by schiessle** [comment]
As you said, it is something different. I don't want to have a way to share a public link by mail, this is already today possible. But to create multiple public links, one for each recipient transparent in the share dialog

**Comment by jancborchardt** [comment]
cc @Rocco83 because this is the »generate unique link per email address so revocation on a per-person basis is possible« feature you are interested in. :)

**Comment by jancborchardt** [comment]
@schiessle any progress on this? Do you need any help or reviews? :)

**Review by nickvergessen — CHANGES_REQUESTED** [review]

**Comment by schiessle** [comment]
I will look if I can add some unit tests, but beside that it is ready to test and review

File: .gitignore [hunk 3]
```diff
@@ -16,6 +16,7 @@
 !/apps/files
 !/apps/federation
 !/apps/federatedfilesharing
+!/apps/sharebymail
```
Comment (nickvergessen) [inline .gitignore]: please alphabetic order

File: apps/sharebymail/appinfo/app.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+$app = new \OCP\AppFramework\App('sharebymail');
```
Comment (nickvergessen) [inline apps/sharebymail/appinfo/app.php]: Is this file really needed?

File: core/shipped.json [hunk 5]
```diff
@@ -34,7 +34,8 @@
     "user_external",
     "user_ldap",
     "user_saml",
-    "workflowengine"
+    "workflowengine",
+    "sharebymail"
```
Comment (nickvergessen) [inline core/shipped.json]: alphabetic order?

File: apps/sharebymail/appinfo/app.php (hunk 4 above)
Comment (schiessle) [inline apps/sharebymail/appinfo/app.php]: no... I removed it

File: .gitignore (hunk 3 above)
Comment (schiessle) [inline .gitignore]: done

File: core/shipped.json (hunk 5 above)
Comment (schiessle) [inline core/shipped.json]: done

**Comment by schiessle** [comment]
Please ignore the failing tests... They are not related to the actual code

@MorrisJobke any idea about this issue for example https://drone.nextcloud.com/nextcloud/server/645/21 can we do anything to make sure that tests don't run out of space?

Please review this... Would be great to get this in soon so that we can move forward. cc @LukasReschke @rullzer @nickvergessen @icewind1991

**Comment by MorrisJobke** [comment]
> @MorrisJobke any idea about this issue for example https://drone.nextcloud.com/nextcloud/server/645/21 can we do anything to make sure that tests don't run out of space?

They actually do, because we have one server with very little space - I rearranged some stuff this evening. I restarted the failed tests.

**Review by LukasReschke — CHANGES_REQUESTED** [review]
👎  until the new files have proper tests in place.

See https://codecov.io/gh/nextcloud/server/compare/f2cae3cee127220a5fde2726af8074837ee81ed1...9dc71f3033a1e2e95cdbec94e14ca05b54a8d800

File: apps/files_sharing/lib/Controller/ShareAPIController.php [hunk 6]
```diff
@@ -196,6 +199,28 @@ protected function formatShare(\OCP\Share\IShare $share, Node $recipientNode = n
...
+	 * Check if one of the users address books knows the exact property, if
+	 * yes we return the full name.
+	 *
+	 * @param string $query
+	 * @param string $property
+	 * @return string
+	 */
+	private function getDisplayNameFromAddressBook($query, $property) {
+		// FIXME: If we inject the contacts manager it gets initialized bofore any address books are registered
+		$result = \OC::$server->getContactsManager()->search($query, [$property]);
```
Comment (rullzer) [inline apps/files_sharing/lib/Controller/ShareAPIController.php]: Mmm we should find a nice solution for that. But for now I guess it is OK

File: apps/files_sharing/lib/Controller/ShareAPIController.php [hunk 7]
```diff
@@ -400,6 +425,17 @@ public function createShare(
 
 			$share->setSharedWith($shareWith);
 			$share->setPermissions($permissions);
+		} else if ($shareType === \OCP\Share::SHARE_TYPE_EMAIL) {
+			if ($share->getNodeType() === 'file') {
```
Comment (rullzer) [inline apps/files_sharing/lib/Controller/ShareAPIController.php]: Note for myself to change this for https://github.com/nextcloud/server/pull/1836

File: apps/files_sharing/lib/Controller/ShareesAPIController.php [hunk 8]
```diff
@@ -584,6 +545,71 @@ protected function searchSharees($search, $itemType, array $shareTypes, $page, $
 	}
 
 	/**
+	 * @param string $search
+	 * @return array
+	 */
+	protected function getEmail($search) {
```
Comment (rullzer) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: Why is this sigular

File: apps/files_sharing/lib/Controller/ShareesAPIController.php [hunk 9]
```diff
@@ -584,6 +545,71 @@ protected function searchSharees($search, $itemType, array $shareTypes, $page, $
...
+			$result['exact'][] = [
+				'label' => $search,
+				'value' => [
+					'shareType' => Share::SHARE_TYPE_EMAIL,
+					'shareWith' => $search,
+				],
+			];
+		}
+
+		$this->reachedEndFor[] = 'emails';
```
Comment (rullzer) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: But this not

File: core/js/sharedialogview.js [hunk 10]
```diff
@@ -154,10 +154,16 @@
 						var users   = result.ocs.data.exact.users.concat(result.ocs.data.users);
 						var groups  = result.ocs.data.exact.groups.concat(result.ocs.data.groups);
 						var remotes = result.ocs.data.exact.remotes.concat(result.ocs.data.remotes);
+						if (typeof(result.ocs.data.emails) !== 'undefined') {
```
Comment (rullzer) [inline core/js/sharedialogview.js]: can this really be undefined?

File: core/js/sharedialogview.js [hunk 11]
```diff
@@ -397,24 +408,24 @@
...
+					tooltip: t('core', 'Share with people on other servers using the syntax username@example.com/nextcloud')
 				});
 			}
 
 			return remoteShareInfo;
 		},
 
 		_renderSharePlaceholderPart: function () {
-			var sharePlaceholder = t('core', 'Share with users…');
+			var sharePlaceholder = t('core', 'Share with users, or by mail...');
```
Comment (rullzer) [inline core/js/sharedialogview.js]: We need to know in js if this is allowed. If the app is not enabled you can't share by mail.

**Comment by rullzer** [comment]
Nice work @schiessle 
Probabaly needs some polishing afterwards but lets take care of that later.

Link was send sucsessfully here. But I'll test in a bit more detail.

Also needs a rebase.

File: core/js/sharedialogview.js (hunk 10 above)
Comment (schiessle) [inline core/js/sharedialogview.js]: it happened to me during testing

**Comment by schiessle** [comment]
> :-1: until the new files have proper tests in place.

@LukasReschke tests are coming in... don't panic :wink: You can still already review and test the code/functionality.

File: apps/files_sharing/lib/Controller/ShareesAPIController.php (hunk 9 above)
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: yeah, this needs to be the key of array that is being sent.

File: core/js/sharedialogview.js (hunk 10 above)
Comment (nickvergessen) [inline core/js/sharedialogview.js]: Then you didn't initialize something correctly...

File: apps/files_sharing/lib/Controller/ShareesAPIController.php [hunk 12]
```diff
@@ -560,13 +504,30 @@ protected function searchSharees($search, $itemType, array $shareTypes, $page, $
...
+
+		// if we have a exact match, either for the federated cloud id or for the
+		// email address we only return the exact match. It is highly unlikely
+		// that the exact same email address and federated cloud id exists
+		if ($mailResults['exactIdMatch'] && !$remoteResults['exactIdMatch']) {
+			$this->result['emails'] = $mailResults['results'];
+			$this->result['exact']['emails'] = $mailResults['exact'];
+		} else if (!$mailResults['exactIdMatch'] && $remoteResults['exactIdMatch']) {
+			$this->result['remotes'] = $remoteResults['results'];
+			$this->result['exact']['remotes'] = $remoteResults['exact'];
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: please define the empty emails array here as well. We always send all keys.

File: apps/files_sharing/lib/Controller/ShareesAPIController.php [hunk 13]
```diff
@@ -560,13 +504,30 @@ protected function searchSharees($search, $itemType, array $shareTypes, $page, $
...
-			$this->getEmails($search);
+			$mailResults = $this->getEmail($search);
+		}
+
+		// if we have a exact match, either for the federated cloud id or for the
+		// email address we only return the exact match. It is highly unlikely
+		// that the exact same email address and federated cloud id exists
+		if ($mailResults['exactIdMatch'] && !$remoteResults['exactIdMatch']) {
+			$this->result['emails'] = $mailResults['results'];
+			$this->result['exact']['emails'] = $mailResults['exact'];
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: please define the empty remotes array here as well. We always send all keys.

File: core/js/sharedialogview.js (hunk 10 above)
Comment (nickvergessen) [inline core/js/sharedialogview.js]: Then you can also kill this

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: apps/files_sharing/lib/Controller/ShareesAPIController.php (hunk 12 above)
Comment (schiessle) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: they are already defined, see: https://github.com/nextcloud/server/blob/9dc71f3033a1e2e95cdbec94e14ca05b54a8d800/apps/files_sharing/lib/Controller/ShareesAPIController.php#L81

File: apps/files_sharing/lib/Controller/ShareesAPIController.php (hunk 13 above)
Comment (schiessle) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: they are already defined, see: https://github.com/nextcloud/server/blob/9dc71f3033a1e2e95cdbec94e14ca05b54a8d800/apps/files_sharing/lib/Controller/ShareesAPIController.php#L81

**Comment by rullzer** [comment]
We need to adjust this provider also for https://github.com/nextcloud/server/pull/339 Depending on which one goes in first.

**Comment by schiessle** [comment]
> We need to adjust this provider also for #339 Depending on which one goes in first.

Done. I also added a lot of unit tests (cc @LukasReschke)... Final review round :wink:

**Comment by rullzer** [comment]
I'd say lets get this in :+1:

**Review by LukasReschke — APPROVED** [review]

**Comment by LukasReschke** [comment]
LGTM

**Comment by LukasReschke** [comment]
@nickvergessen request changes are pending here.

**Review by nickvergessen — APPROVED** [review]

**Comment by nickvergessen** [comment]
Yeah lets do that...

**Pull request merged**